covid_moonshot/analysis/diffnet.py

```diff
@@ -0,0 +1,151 @@
+                           complex_overlap=result['analysis']['complex_phase']['free_energy']['bar_overlap'],
+                           solvent_overlap=result['analysis']['solvent_phase']['free_energy']['bar_overlap'])
+
+        # see if either ligand has an experimental affinity associated node
+        for state in ['start', 'end']:
+            if f'{state}_pIC50' in result['details'].keys():
+                for graph in graphs.values():
+                    pIC50 = float(result['details'][f'{state}_pIC50'])
+                    title = result['details'][f'{state}_title']
+                    if graph.has_node(result['details'][f'{state}_title']):
+                        # see if it already has an pIC50
+                        if 'pIC50' in graph.nodes[title]:
+                            assert(pIC50 == graph.nodes[title]['pIC50']), \
+                                   f"Trying to add an pIC50, {float(result['details'][f'{state}_pIC50'])}, \
+                                   that disagrees with a previous value for \
+                                   node {graph.nodes[title]}"
+                        else:
+                            # or add the pIC50 if it's not available
+                            graph.nodes[title]['pIC50'] = pIC50
+                            ic50 = 10**(-pIC50)
+                            dg = _ic50_to_dG(ic50)
+                            graph.nodes[title]['exp_DG'] = dg
+                            exp_dDG = _ic50_to_dG(10**(-EXPERIMENTAL_ERROR))
+                            graph.nodes[title]['exp_dDG'] = exp_dDG
+
+    # TODO this can be wrapped in a function
+    _logger.info(f'There are {len(graphs)} unique protein files used')
+    graphs = separate_graphs(graphs)
+    _logger.info(f'There are {len(graphs)} graphs \
+    after splitting into weakly connected sets')
+
+    # now combine the relative simulations to get a per-ligand estimate
+    for name, graph in graphs.items():
+        f_i, C = stats.mle(graph, factor='f_ij', node_factor='exp_DG')
+        errs = np.diag(C)
+
+        exp_values = [n[1]['exp_DG'] for n in graph.nodes(data=True)
+                      if 'exp_DG' in n[1]]
+        if len(exp_values) == 0:
+            _logger.warning(f'No experimental value in graph {name}. \
+            Results are only useful for relative comparisions \
+            within this set, NOT with other sets and should not \
+            be considered as absolute predictions')
+            for dg, dg_err, node in zip(f_i, errs, graph.nodes(data=True)):
+                node[1]['rel_f_i'] = dg
+                node[1]['rel_df_i'] = dg_err
+        else:
+            for dg, dg_err, node in zip(f_i, errs, graph.nodes(data=True)):
+                # shift the calc RBFE to the known experimental values
+                node[1]['f_i'] = dg
+                node[1]['df_i'] = dg_err
+
+    # not sure how to best return from this script
+    return graphs
```

Comment:
I think we'll want to return both the graph (for plotting purposes) and updated key entries to a `.json` for easy manipulation later.

Reply:
(or a "`.json` like" object e.g a `list` or `dict`)

Reply:
Agreed that it would be useful to return both the graph and some json-compatible structure for inclusion in the analysis.json!

For the latter, I've been following the convention of describing the structure as a dataclass, e.g. like https://github.com/choderalab/covid-moonshot-infra/blob/266612d6ceef234ad6b77840f7e23cb65700dc64/covid_moonshot/core.py#L34-L40. 

This is nice as a form of documentation and also allows using something like [mypy](http://mypy-lang.org/) to check that we're not accessing fields incorrectly in downstream code.